### PR TITLE
Add ONNX Training Post-Passes to Front-End - Cont

### PR DESCRIPTION
**Pull request by BowenBao**
Replacing #3942

Description: This PR adds ONNX post-process to the PyTorch front-end. The User can disable these passes by setting _enable_internal_postprocess to False (default is True).

Motivation and Context

    These are common training post-passes, by ading them to the front-end, we abstract complexity from the user.

### CODE REVIEW

**Comment by Tixxx** [comment]
I don't think I agree with the problem statement here. This is essentially moving some transformations some of which were meant to be model-by-model hacks and workarounds from user code base to ORTTrainer for generic use. Some of them are not needed for some models. I.E. I'm not sure why we still have layerNorm when we already perform training-compatible layernorm fusion in the backend. Also I think the nll_loss fusion should really be a back-end transformation pass.

File: orttraining/orttraining/python/ort_trainer.py [hunk 1]
```diff
@@ -538,6 +483,7 @@ def __init__(self, model, loss_fn, model_desc, training_optimizer_name, map_opti
             use_mixed_precision: flag to enable mixed precision (aka fp16) (default is False)
             allreduce_post_accumulation: controls whether overlaping gradient computation with allreduce (default is False)
             partition_optimizer: controls whether to partition the optimizer state (default is False)
+            _enable_internal_postprocess: whether to run or not the internal postprocesses (default is True)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:486]: Replace `postprocess_model` by `_extra_postprocess` on line 480

File: orttraining/orttraining/python/postprocess.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
+import sys
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/postprocess.py:1]: Lara's PR has 259 lines for postprocess.py. This has  332 lines. What is the difference?

File: orttraining/orttraining/test/python/onnxruntime_test_postprocess.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+        onnx_model = self.get_onnx_model(model, model_desc, input_args, device)
+
+        count_layer_norm = self.count_nodes(onnx_model, "LayerNormalization")
+        count_nodes = self.count_all_nodes(onnx_model)
+       
+        assert count_layer_norm == 1
+        assert count_nodes == 3
+
+    def test_expand(self):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/onnxruntime_test_postprocess.py:105]: Did you fix the issue Lara has reported on the original PR:?

her words:

"In this test I the learning rate as the last input in input_args with the descriptor " IODescription('Learning_Rate', [1,], torch.float32)". But I still get a binding 'Learning_Rate' error. I follow the same logic in test_layer_norm an don't get the error. Any idea why this is happening?"

File: orttraining/orttraining/python/ort_trainer.py [hunk 4]
```diff
@@ -321,7 +265,7 @@ def convert_model_loss_fn_to_onnx(model, loss_fn, model_desc, device, inputs, op
                        output_names=output_names,
                        opset_version=opset_version,
                        dynamic_axes=dynamic_axes,
-                       training=True,
+                       training=torch.onnx.TrainingMode.TRAINING,
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:268]: Are we supporting older pytorch? This will not work and will lead to silent errors. On my tests, I forgot to change that and the script run and the numbers didnt match. Super tricky issue as it doesn't raise any exception

File: orttraining/orttraining/test/python/onnxruntime_test_postprocess.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+        device = torch.device("cpu")
+        target = torch.ones(20, 10, 10, dtype=torch.int64).to(device)
+        model = LayerNormNet(target)
+        input = torch.randn(20, 5, 10, 10, dtype=torch.float32).to(device)
+
+        input_desc = IODescription('input', [], "float32")
+        output0_desc = IODescription('output0', [], "float32")
+        output1_desc = IODescription('output1', [20, 5, 10, 10], "float32")
+        output2_desc = IODescription('output2', [20, 5, 10, 10], "float32")
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/onnxruntime_test_postprocess.py:91]: nit: unused `output2_desc`?

**Comment by thiagocrepaldi** [comment]
This PR depends on #4052 to allow training with Dropout opset 12

File: orttraining/orttraining/python/ort_trainer.py (hunk 4 above)
Comment (BowenBao) [reply orttraining/orttraining/python/ort_trainer.py:268]: exactly, this will be fixed similar to `enable_onnx_checker` flag depending on pytorch version.

File: orttraining/orttraining/python/postprocess.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+from onnx import *
+import onnx
+import numpy as np
+import struct
+
+from onnx import helper
+from onnx import numpy_helper
+
+
+def run_postprocess(model):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply orttraining/orttraining/python/postprocess.py:12]: correct, backend is handling this pass

File: orttraining/orttraining/test/python/onnxruntime_test_postprocess.py (hunk 3 above)
Comment (BowenBao) [reply orttraining/orttraining/test/python/onnxruntime_test_postprocess.py:108]: yes, the learning rate error is resolved. The error was due to that the model has no weights to train. Adding a linear layer in the model resolves the issue.

File: orttraining/orttraining/python/postprocess.py (hunk 2 above)
Comment (BowenBao) [reply orttraining/orttraining/python/postprocess.py:1]: rebase error that includes additional lines that are removed later. fixed

File: orttraining/orttraining/test/python/onnxruntime_test_postprocess.py (hunk 3 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/test/python/onnxruntime_test_postprocess.py:108]: Should we check if at least one weight is present in the model on ORTTrainer? That would make the API more robust

File: orttraining/orttraining/python/ort_trainer.py [hunk 7]
```diff
@@ -311,17 +255,20 @@ def convert_model_loss_fn_to_onnx(model, loss_fn, model_desc, device, inputs, op
...
+    # This option was added after 1.6 release.
+    if LooseVersion(torch.__version__) >= LooseVersion('1.6.0'):
+        other_export_options['training'] = torch.onnx.TrainingMode.TRAINING
 
     torch.onnx._export(model, tuple(sample_inputs), f,
                        input_names=input_names,
                        output_names=output_names,
                        opset_version=opset_version,
                        dynamic_axes=dynamic_axes,
-                       training=True,
```
Comment (suffiank) [inline orttraining/orttraining/python/ort_trainer.py:324]: Is there a PR open on pytorch so it will treat training=True as if training=torch.onnx.TrainingMode.TRAINING? I believe current nightly will return evaluation graph despite training=True.

Comment (BowenBao) [reply orttraining/orttraining/python/ort_trainer.py:324]: training becomes an enum instead of bool in pytorch nightly. True is not a valid input argument in that sense. This PR also updates this parameter in case PyTorch version is >= 1.6

File: orttraining/orttraining/python/postprocess.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    attr.name = 'value'
+    if t_value is not None:
+        attr.type = 4
+        attr.t.CopyFrom(t_value)
+    else:
+        attr.type = 1
+        attr.f = f_value
+    return const_node
+
+def layer_norm_transform(model):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply orttraining/orttraining/python/postprocess.py:209]: done. Difference is that updated version takes care of possible Cast nodes around Pow node, which is introduced in newer version of pytorch.

File: orttraining/orttraining/python/postprocess.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    for node in layer_norm_nodes:
+        all_nodes.append(node)
+
+    graph.ClearField("node")
+    graph.node.extend(all_nodes)
+    return model
+
+# Fuse SoftmaxCrossEntropy
+
+def fuse_softmaxNLL_to_softmaxCE(onnx_model):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply orttraining/orttraining/python/postprocess.py:384]: done.

File: orttraining/orttraining/python/postprocess.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    for node in model.graph.node:
+        for input in node.input:
+            if input == arg:
+                result.append(node)
+    return result[0] if len(result) == 1 else result
+
+
+# Expand Shape PostProcess
+
+def fix_expand_shape(model):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply orttraining/orttraining/python/postprocess.py:50]: done.

**Comment by BowenBao** [comment]
CI failing due to a recently introduced bug in testing, the bug is fixed in #4061
